src/Ltb/Ldap.php

```diff
@@ -136,7 +136,147 @@ static function sorted_search($ldap, $ldap_base, $ldap_filter, $attributes, $sor
+     * @param \LDAP\Connection|array $ldap An LDAP\Connection instance, returned by ldap_connect()
+     * @param string $dn the dn of the user
+     * @param string $oldpassword the old password
+     * @param string $password the new password
+     * @return array [$error_code, $error_msg]
+     */
+    static function change_ad_password_as_user($ldap, $dn, $oldpassword, $password): array {
+        # The AD password change procedure is modifying the attribute unicodePwd by
+        # first deleting unicodePwd with the old password and them adding it with the
+        # the new password
+        $oldpassword_hashed = make_ad_password($oldpassword);
+
+        $modifications = array(
+            array(
+                "attrib" => "unicodePwd",
+                "modtype" => LDAP_MODIFY_BATCH_REMOVE,
+                "values" => array($oldpassword_hashed),
+            ),
+            array(
+                "attrib" => "unicodePwd",
+                "modtype" => LDAP_MODIFY_BATCH_ADD,
+                "values" => array($password),
+            ),
+        );
+
+        ldap_modify_batch($ldap, $dn, $modifications);
+        $error_code = ldap_errno($ldap);
+        $error_msg = ldap_error($ldap);
+        return array($error_code, $error_msg);
+    }
+    
+    /**
+     * Changes the Password using extended password modification
+     * @param \LDAP\Connection|array $ldap An LDAP\Connection instance, returned by ldap_connect()
+     * @param string $dn the dn of the user
+     * @param string $oldpassword the old password
+     * @param string $password the new password
+     * @param array $userdata
+     * @param bool $use_ppolicy_control
+     * @return array [$error_code, $error_msg, $ppolicy_error_code]
+     */
+    static function change_password_with_exop($ldap, $dn, $oldpassword, $password, $userdata, $use_ppolicy_control): array {
+        $ppolicy_error_code = "";
+        $exop_passwd = FALSE;
+        if ( $use_ppolicy_control ) {
+            $ctrls = array();
+            $exop_passwd = ldap_exop_passwd($ldap, $dn, $oldpassword, $password, $ctrls);
+            $error_code = ldap_errno($ldap);
+            $error_msg = ldap_error($ldap);
+            if (!$exop_passwd) {
+                if (isset($ctrls[LDAP_CONTROL_PASSWORDPOLICYRESPONSE])) {
+                    $value = $ctrls[LDAP_CONTROL_PASSWORDPOLICYRESPONSE]['value'];
+                    if (isset($value['error'])) {
+                        $ppolicy_error_code = $value['error'];
+                        error_log("LDAP - Ppolicy error code: $ppolicy_error_code");
+                    }
+                }
+            }
+        } else {
+            $exop_passwd = ldap_exop_passwd($ldap, $dn, $oldpassword, $password);
+            $error_code = ldap_errno($ldap);
+            $error_msg = ldap_error($ldap);
+        }
+        if ($exop_passwd === TRUE) {
+            # If password change works update other data
+            if (!empty($userdata)) {
```

Comment:
So this function does not only change the password, it also modify some attributes defined in $userdata?

It think it is preferable to do atomic actions only. If the calling program needs to modify the password + some other attribute, it should call two different functions.

Reply:
This comes from the fact that in SSP, you can also set some shadow, samba and ad data, see the appropriate functions `set_xxx_data` in `Password.php`. Of course, it can also be set in an additional ldap replace command, but I think one replace task is more appropriate.